### PR TITLE
[15.0][MIG] web_widget_domain_editor_dialog

### PR DESCRIPTION
**Pull request by celm1990**

### CODE REVIEW

**Comment by celm1990** [comment]
ping @OCA/web-maintainers @OCA/oca-contributors

File: web_widget_domain_editor_dialog/__manifest__.py [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    "category": "Web",
+    "author": "Tecnativa," "Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/web",
+    "license": "AGPL-3",
+    "depends": ["web"],
+    "assets": {
+        "web.assets_backend": [
+            "/web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js",
+            "/web_widget_domain_editor_dialog/static/src/js/basic_fields.js",
+        ],
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/__manifest__.py]:
```suggestion
        "web.assets_backend": ["/web_widget_domain_editor_dialog/static/src/js/*.js"],
```

File: web_widget_domain_editor_dialog/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    },
+    "data": [],
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/__manifest__.py]:
```suggestion
```

File: web_widget_domain_editor_dialog/static/src/js/basic_fields.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
+/* Copyright 2019 Tecnativa - David Vidal
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). */
+
+odoo.define("web_widget_domain_editor_dialog.basic_fields", function (require) {
+    "use strict";
+
+    const core = require("web.core");
+    const basic_fields = require("web.basic_fields");
+    const DomainEditorDialog = require("web_widget_domain_editor_dialog.DomainEditorDialog");
+    const _t = core._t;
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/basic_fields.js]:
```suggestion
    const {_t} = require("web.core");
    const {FieldDomain} = require("web.basic_fields");
    const DomainEditorDialog = require("web_widget_domain_editor_dialog.DomainEditorDialog");
```

File: web_widget_domain_editor_dialog/static/src/js/basic_fields.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    basic_fields.FieldDomain.include({
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/basic_fields.js]:
```suggestion
    FieldDomain.include({
```

File: web_widget_domain_editor_dialog/static/src/js/basic_fields.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+                no_create: true,
+                on_selected: function (selected_ids) {
+                    _this.domainSelector
+                        .setDomain(this.get_domain(selected_ids))
+                        .then(_this._replaceContent.bind(_this));
+                    _this.trigger_up("domain_changed", {
+                        child: _this,
+                        alreadyRedrawn: true,
+                    });
+                },
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/basic_fields.js]: Callback function would be way cooler

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+/* Copyright 2019 Tecnativa - David Vidal
+ * License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl). */
+
+odoo.define("web_widget_domain_editor_dialog.DomainEditorDialog", function (require) {
+    "use strict";
+
+    const core = require("web.core");
+    const view_dialogs = require("web.view_dialogs");
+    const Domain = require("web.Domain");
+    const _t = core._t;
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js]:
```suggestion
    const {_t} = require("web.core");
    const {SelectCreateDialog} = require("web.view_dialogs");
    const Domain = require("web.Domain");
```

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+
+    const DomainEditorDialog = view_dialogs.SelectCreateDialog.extend({
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js]:
```suggestion
    const DomainEditorDialog = SelectCreateDialog.extend({
```

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+                dynamicFilters: [
+                    {
+                        description: _.str.sprintf(_t("Selected domain")),
+                        domain: Domain.prototype.stringToArray(
+                            _this.options.default_domain
+                        ),
+                    },
+                ],
+            });
+        },
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js]:
```suggestion
        init: function () {
            this._super.apply(this, arguments);
            this.options = _.defaults(this.options, {
                dynamicFilters: [
                    {
                        description: _t("Selected domain"),
                        domain: Domain.stringToArray(this.options.default_domain),
                    },
                ],
            });
        },
```
Pretty sure that should work just as well

**Review by chienandalu — COMMENTED** [review]
It isn't working for me. Tested in the domain widget of the filters form:

- I edit a domain in the dialog
- The domain doesn't do it into the domain field:

image

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+                    ).concat(group_domain);
+                }
+            } else {
+                const ids = selected_ids.map((x) => {
+                    return x.id;
+                });
+                domain = domain.concat([["id", "in", ids]]);
+            }
+            return domain.concat(group_domain);
+        },
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js:73]: You could do something like
```suggestion
    get_domain: function (selected_ids) {
        const search_data = this.viewController.renderer.state;
        const domain = search_data.domain;
        if (
            this.$(".o_list_record_selector input").prop("checked")
            && search_data.groupedBy.length
        ) {
            const group_domain = search_data.data
                .filter((x) => x.res_ids.length)
                .map((x) => x.domain)
                .flat()
                .filter((x) => !domain.some((d) => _.isEqual(x, d)))
                .filter((x) => x != "&");
            if (group_domain.length > 1) {
                group_domain.unshift(Array(group_domain.length - 1).fill("|"));
            }
            return domain.concat(group_domain);
        }
        return domain.concat([["id", "in", selected_ids.map((x) => x.id)]]);
    },
```
this.

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+        on_view_list_loaded: () => {
+            this.$(".o_list_record_selector input").prop("checked", true);
+            this.$footer
+                .find(".o_selectcreatepopup_search_select")
+                .prop("disabled", false);
+        },
```
Comment (Du-ma) [inline web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js]:
```suggestion
        on_view_list_loaded: function () {
            this.$(".o_list_record_selector input").prop("checked", true);
            this.$footer
                .find(".o_selectcreatepopup_search_select")
                .prop("disabled", false);
        },
```
This should be a normal function, try it out.

**Comment by celm1990** [comment]
> It isn't working for me. Tested in the domain widget of the filters form:
> 
> * I edit a domain in the dialog
> * The domain doesn't do it into the domain field:
> 
> image

@chienandalu  I fixed this, please can try again please?

**Review by chienandalu — COMMENTED** [review]
Now the first filter works fine, but when I go back and try to edit it again, I get this traceback:

<details>
<summary>Traceback Error</summary>
<pre>
UncaughtPromiseError
Uncaught Promise > Control panel model extension failed to evaluate domain:/n{}

Error: Control panel model extension failed to evaluate domain:/n{}
    at ControlPanelModelExtension.getDomain (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:44300:23) (/web/static/src/legacy/js/control_panel/control_panel_model_extension.js:433)
    at ControlPanelModelExtension.get (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:44059:44) (/web/static/src/legacy/js/control_panel/control_panel_model_extension.js:192)
    at ActionModel.__get (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:93242:46) (/web/static/src/legacy/js/model.js:399)
    at ActionModel.__get (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:63027:35) (/web/static/src/legacy/js/views/action_model.js:102)
    at ActionModel._getQuery (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:63087:34) (/web/static/src/legacy/js/views/action_model.js:162)
    at ActionModel.get (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:62995:70) (/web/static/src/legacy/js/views/action_model.js:70)
    at Class.getController (http://oca-web-15-0-pr2129-070f1871ef3a.runboat.odoo-community.org/web/assets/debug/web.assets_backend.js:62782:43) (/web/static/src/legacy/js/views/abstract_view.js:314)
</pre>
</details>

Steps:

Peek 27-01-2022 08-54

**Comment by celm1990** [comment]
> Now the first filter works fine, but when I go back and try to edit it again, I get this traceback:

It is rare that with certain filters it works and with others it does not. I'll be checking it

**Comment by celm1990** [comment]
@chienandalu i tested on runbot V14/V13 and get same error, and has not found solution, can you help me please??
Note: I see error happend with filters with fields type `boolean `

`[["is_company","=",True]]`

Runbot V14(same error as V15)
image
Runbot V13
image



URL Runbot instance V14:
http://3528491-14-0-5ea79d.runbot1-3.odoo-community.org/web?debug=assets#id=1&model=ir.filters&view_type=form&cids=1&menu_id=95

**Comment by chienandalu** [comment]
Oh, I thought I had this fix merged already: https://github.com/OCA/web/pull/2098

@pedrobaeza can we please?

**Comment by pedrobaeza** [comment]
Merged, please cherry-pick it here

**Comment by chienandalu** [comment]
Thanks :)

**Comment by celm1990** [comment]
@chienandalu @pedrobaeza I cherry-pick and now working fine, please update your reviews

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js (hunk 9 above)
Comment (celm1990) [reply web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js:73]: @Du-ma I tested your suggestion but domain is more complex, please see my video

https://user-images.githubusercontent.com/7775116/151639804-493765ed-fcae-4dd8-a650-2bdb794f6aa1.mp4

**Comment by celm1990** [comment]
ping @chienandalu @pedrobaeza

File: web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js (hunk 9 above)
Comment (Du-ma) [reply web_widget_domain_editor_dialog/static/src/js/widget_domain_editor_dialog.js:73]: Sure, ill try refactoring it again when I have a free moment :grin:
Have a feeling that this is just way too over complicated than it needs to be.

**Review by chienandalu — APPROVED** [review]
Thanks @celm1990 :slightly_smiling_face: :+1:

**Pull request merged**